share/man/man1/appjail.1
.\"Copyright (c) 2024-2026, Jesús Daniel Colmenares Oviedo <DtxdF@disroot.org>
.\"All rights reserved.
.\"
.\"Redistribution and use in source and binary forms, with or without
.\"modification, are permitted provided that the following conditions are met:
.\"
.\"* Redistributions of source code must retain the above copyright notice, this
.\"  list of conditions and the following disclaimer.
.\"
.\"* Redistributions in binary form must reproduce the above copyright notice,
.\"  this list of conditions and the following disclaimer in the documentation
.\"  and/or other materials provided with the distribution.
.\"
.\"* Neither the name of the copyright holder nor the names of its
.\"  contributors may be used to endorse or promote products derived from
.\"  this software without specific prior written permission.
.\"
.\"THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
.\"AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\"IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
.\"DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
.\"FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\"DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
.\"SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
.\"CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
.\"OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
.\"OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.Dd September 08, 2026
.Dt APPJAIL 1
.Os
.Sh NAME
.Nm appjail
.Nd Simple and easy-to-use tool for creating portable jails
.Sh SYNOPSIS
.Nm appjail
.Ar command
.Op Ar args Ns " " Ns "..."
.Sh DESCRIPTION
AppJail is an open-source BSD-3 licensed framework entirely written in POSIX shell
and C to create isolated, portable and easy to deploy environments using FreeBSD
jails that behaves like an application.
.Pp
See
.Xr appjail-tutorial 7
to get started and
.Lk https://appjail.readthedocs.io/en/latest/getting-started/ "Getting Started"
if you prefer an online guide.
.Xr appjail-tutorial 7
provides a quick guide to configuring AppJail, RC scripts, the packet filter,
and some recommendations.
.Sy This tutorial is not suitable for all kind of users ","
but for beginners or users who do not need all the features of AppJail.
.Sh SUBCOMMANDS
.Ss Tutorial
.Bl -tag -width xxx
.It Xr appjail-tutorial 7
A tutorial introduction to AppJail.
.El
.Ss Configuration
.Bl -tag -width xxx
.It Xr appjail.conf 5
Configuration parameters for AppJail.
.El
.Ss Image
.Bl -tag -width xxx
.It Xr appjail-ajspec 5
AJSPEC file specification.
.It Xr appjail-image 1
Create, list, remove, update images and much more.
.El
.Ss Automation
.Bl -tag -width xxx
.It Xr appjail-initscript 5
Jail automation script.
.It Xr appjail-makejail 5
Build a jail using a Makejail file.
.It Xr appjail-label 1
Add metadata to jails.
.El
.Ss Templates
.Bl -tag -width xxx
.It Xr appjail-config 1
Configuration utility for templates.
.It Xr appjail-template 5
Template file specification.
.El
.Ss Makejails
.Bl -tag -width xxx
.It Xr appjail-apply 1
Apply a Makejail to an existing jail.
.It Xr appjail-makejail 1
Build a jail using a Makejail file.
.El
.Ss Networking
.Bl -tag -width xxx
.It Xr appjail-network
Create, remove or manage virtual networks for jails.
.It Xr appjail-nat 1
Mapping local IPv4 address to an external IPv4 address.
.It Xr appjail-expose 1
Port forwarding from host port to jail port.
.El
.Ss Help
.Bl -tag -width xxx
.It Xr appjail-help 1
Display help information about AppJail.
.It Xr appjail-usage 1
Display the syntax of a given command.
.It Xr appjail-version 1
Display version information about AppJail.
.El
.Ss DNS
.Bl -tag -width xxx
.It Xr appjail-dns 8
DNS information for AppJail users.
.El
.Ss X11
.Bl -tag -width xxx
.It Xr appjail-x11 1
Execute x11 applications in jails.
.El
.Ss ZFS
.Bl -tag -width xxx
.It Xr appjail-zfs 1
Common ZFS commands for jails.
.El
.Ss Source
.Bl -tag -width xxx
.It Xr appjail-deleteOld 1
Run delete-old and delete-old-libs targets on a given release or jail.
.It Xr appjail-checkOld 1
Run check-old target on a given release or jail
.It Xr appjail-etcupdate 1
Manage updates to system files not updated by installworld.
.El
.Ss Startup Management
.Bl -tag -width xxx
.It Xr appjail-restart 1
Restart a running jail.
.It Xr appjail-rstop 1
Stop a jail and the jails it depends on.
.It Xr appjail-stop 1
Stop a running jail.
.It Xr appjail-startup 1
Unattended command to start jails and other related things.
.It Xr appjail-start 1
Start a stopped jail.
.It Xr appjail-status 1
Show the status of a jail.
.El
.Ss Volume Management
.Bl -tag -width xxx
.It Xr appjail-fstab 1
Static information about the file systems of a jail.
.It Xr appjail-volume 1
Volume management for jails.
.El
.Ss Resource Management
.Bl -tag -width xxx
.It Xr appjail-limits 1
Display and update resource limits database for jails.
.It Xr appjail-cpuset 1
Configure processor sets for jails.
.El
.Ss Update / Upgrade
.Bl -tag -width xxx
.It Xr appjail-update 1
Update a thick jail or a release.
.It Xr appjail-upgrade 1
Upgrade a jail or a release.
.El
.Ss Argument Persistence
.Bl -tag -width xxx
.It Xr appjail-disable 1
Remove default parameters for stages.
.It Xr appjail-enable 1
Set default parameters for stages.
.It Xr appjail-enabled 1
Get the value for a given enabled parameter.
.El
.Ss Supervisor
.Bl -tag -width xxx
.It Xr appjail-healthcheck 1
Keep your jails healthy and running.
.El
.Ss Logs Management
.Bl -tag -width xxx
.It Xr appjail-logs 1
Log management for jails.
.El
.Ss Xr rc 8 Management
.Bl -tag -width xxx
.It Xr appjail-service 1
Control (start/stop/etc.) or list system services within a jail.
.It Xr appjail-sysrc 1
Safely edit system rc files within a jail.
.El
.Ss Trusted Users
.Bl -tag -width xxx
.It Xr appjail-user 8
Hints and parameters for trusted users.
.El
.Ss Jail Creation
.Bl -tag -width xxx
.It Xr appjail-jail 1
Creates, removes, lists and configures jails.
.It Xr appjail-quick 1
Create a pre-configured jail.
.El
.Ss Command Execution
.Bl -tag -width xxx
.It Xr appjail-cmd 1
Execute commands in a jail.
.El
.Ss Dynamic Device Management
.Bl -tag -width xxx
.It Xr appjail-devfs 1
Dynamic DEVFS ruleset management.
.El
.Ss Package Management
.Bl -tag -width xxx
.It Xr appjail-pkg 1
Manipulate packages within a jail.
.El
.Ss Release Creation
.Bl -tag -width xxx
.It Xr appjail-fetch 1
Fetch and extract components to create jails.
.El
.Ss Ephemeral Concept
.Bl -tag -width xxx
.It Xr appjail-ephemeral 7
cattle, pets and jails.
.El
.Ss Login
.Bl -tag -width xxx
.It Xr appjail-login 1
Log into the jail.
.El
.Ss User-defined Stages
.Bl -tag -width xxx
.It Xr appjail-run 1
Run user-defined stages.
.El
.Ss OCI
.Bl -tag -width xxx
.It Xr appjail-oci 1
Commands for interpreting OCI images.
.Ss SECRETS
.Bl -tag -width xxx
.It Xr appjail-secrets 1
Secrets management for jails.
.El
.El
.Sh SIGNALS
AppJail returns
.Sy 70
when a signal is captured or other depending on the last command executed and
interrupted. The following signals are captured:
.Pp
.Bl -tag -compact -width xx
.It Sy SIGHUP
.It Sy SIGINT
.It Sy SIGQUIT
.It Sy SIGTERM
.It Sy SIGXCPU
.It Sy SIGXFSZ
.El
.Pp
The following signals are ignored:
.Pp
.Bl -tag -compact -width xx
.It Sy SIGALRM
.It Sy SIGVTALRM
.It Sy SIGPROF
.It Sy SIGUSR1
.It Sy SIGUSR2
.El
.Pp
If a signal is captured or simply terminates execution, AppJail performs several
operations that depend entirely on the subcommand, but some are general: send a
.Sy SIGTERM
.Pq or Sy SIGKILL No if Sy SIGTERM No has no effect after a while
to all child processes created by the main process and delete temporary files.
.Sh EXIT STATUS
.Ex -std
It may fail for one of the following reasons:
.Pp
.Bl -dash -compact
.It
.Em EX_USAGE ":"
The
.Ar command
was not specified.
.It
.Em EX_NOINPUT ":"
The
.Ar command
does not exist.
.It
.Em EX_SOFTWARE ":"
The function responsible for the implementation of the
.Ar command
does not exist.
.It
.Em EX_IOERR ":"
Error creating a ZFS dataset or a directory.
.It
.Em EX_DATAERR ":"
The log name used to record the AppJail session is invalid.
.El
.Sh SEE ALSO
.Xr appjail-tutorial 7
.Xr appjail-help 1
.Xr appjail-usage 1
.Xr sysexits 3
.Sh AUTHORS
.An Jesús Daniel Colmenares Oviedo Aq Mt DtxdF@disroot.org


appjail.sh
#!/bin/sh
#
# Copyright (c) 2022-2026, Jesús Daniel Colmenares Oviedo <DtxdF@disroot.org>
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
# * Redistributions of source code must retain the above copyright notice, this
#   list of conditions and the following disclaimer.
#
# * Redistributions in binary form must reproduce the above copyright notice,
#   this list of conditions and the following disclaimer in the documentation
#   and/or other materials provided with the distribution.
#
# * Neither the name of the copyright holder nor the names of its
#   contributors may be used to endorse or promote products derived from
#   this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

APPJAIL_PREFIX="${APPJAIL_PREFIX:-%%PREFIX%%}"
APPJAIL_PROGRAM="${APPJAIL_PREFIX}/bin/appjail"
CONFIG="${APPJAIL_PREFIX}/share/appjail/files/config.conf"

main()
{
	set -T

	. "${CONFIG}"

	if [ `id -u` -ne 0 ]; then
		exec "${SCRIPTSDIR}/ajuser.sh" "$@"
	fi

	. "${LIBDIR}/load"

	# For convenience, these will be loaded.
	lib_load "${LIBDIR}/sysexits"
	lib_load "${LIBDIR}/atexit"
	lib_load "${LIBDIR}/log"
	lib_load "${LIBDIR}/check_func"
	lib_load "${LIBDIR}/zfs"

	# Create the data directory for functions that use it at startup.
	if ! lib_zfs_mkroot; then
		lib_err ${EX_IOERR} "Error creating ${DATADIR}"
	fi

	lib_init_logtime

	local cmd=$1; shift
	if [ -z "${cmd}" ]; then
		usage
	fi
	
	if [ ! -x "${COMMANDS}/${cmd}" ]; then
		lib_err ${EX_NOINPUT} "Command \"${cmd}\" does not exist."
	fi

	lib_load "${COMMANDS}/${cmd}"
	
	# logging
	if [ "${ENABLE_LOGGING_OUTPUT}" != "0" -a -z "${AJ_LOG_SESSION_ID}" ]; then
		lib_debug "Running: ${SESSION_ID_NAME}"

		local errlevel

		local session_id
		session_id=`sh -c "${SESSION_ID_NAME}"`

		errlevel=$?
		if [ ${errlevel} -ne 0 ]; then
			lib_err ${errlevel} "{SESSION_ID_NAME} exits with a non-zero exit status."
		fi

		if lib_check_ispath "${session_id}"; then
			lib_err ${EX_DATAERR} -- "${session_id}: invalid log name."
		fi

		lib_zfs_mklogdir "commands" "${cmd}" "output"

		exec env \
			AJ_LOG_SESSION_ID="${session_id}" \
			script -a -t "${SCRIPT_TIME}" -- \
				"${LOGDIR}/commands/${cmd}/output/${session_id}" \
				"${APPJAIL_PROGRAM}" \
				"${cmd}" \
				"$@"

		exit 0
	fi

	if ! lib_check_func "${cmd}_main"; then
		lib_err ${EX_SOFTWARE} "${cmd}_main function does not exist in the ${cmd} command."
	fi

	lib_atexit_init

	if [ -z "${APPJAIL_KILL_TREE}" ]; then
		# To avoid using kill_tree.sh more times than necessary.
		export APPJAIL_KILL_TREE=1

		local kill_tree_cmd
		kill_tree_cmd=`lib_escape_string "${SCRIPTSDIR}/kill_tree.sh"`

		local config_file
		config_file=`lib_escape_string "${CONFIG}"`

		lib_atexit_add "\"${kill_tree_cmd}\" -c \"${config_file}\" -p $$ > /dev/null 2>&1"
	fi

	if [ -n "${HOOKSDIR}" ] && [ -d "${HOOKSDIR}/pre.d" ]; then
		for hook in "${HOOKSDIR}/pre.d"/*; do
			if [ "${hook}" = "${HOOKSDIR}/pre.d/*" ]; then
				break
			fi

			if [ ! -x "${hook}" ]; then
				lib_debug "Missing execution bit permission for hook '${hook}'"
				continue
			fi

			local errlevel

			env \
				APPJAIL_CONFIG="${CONFIG}" \
				"${hook}" "${cmd}" "$@"

			errlevel=$?

			if [ ${errlevel} -ne 0 ]; then
				lib_err ${errlevel} "Hook '${hook}' exits with a non-zero exit status (pre)"
			fi
		done
	fi

	${cmd}_main "$@" || return $?

	if [ -d "${HOOKSDIR}" ]; then
		for hook in "${HOOKSDIR}/post.d"/*; do
			if [ "${hook}" = "${HOOKSDIR}/post.d/*" ]; then
				break
			fi

			if [ ! -x "${hook}" ]; then
				lib_debug "Missing execution bit permission for hook '${hook}'"
				continue
			fi

			local errlevel

			env \
				APPJAIL_CONFIG="${CONFIG}" \
				"${hook}" "${cmd}" "$@"

			errlevel=$?

			if [ ${errlevel} -ne 0 ]; then
				lib_err ${errlevel} "Hook '${hook}' exits with a non-zero exit status (post)"
			fi
		done
	fi
}

usage()
{
	echo "usage: appjail <command> [<args> ...]" >&2
	exit 64 # EX_USAGE
}

main "$@"
